File: Core/Repository/Repository.swift
import Foundation

public protocol UserRepository: Sendable {
    func currentUser() async throws -> User?
    func availableUsers() async throws -> [User]
    func setCurrentUser(id: EntityID) async throws
    func user(id: EntityID) async throws -> User?
    func users(role: Role?) async throws -> [User]
    func createAccount(email: String, password: String, fullName: String, fullNameAr: String, role: Role, branchID: EntityID?) async throws
    func linkChild(userID: EntityID, athleteID: EntityID) async throws
    func unlinkChild(userID: EntityID, athleteID: EntityID) async throws
}

public protocol BranchRepository: Sendable {
    func branches() async throws -> [Branch]
    func branch(id: EntityID) async throws -> Branch?
}

public protocol AthleteRepository: Sendable {
    func athletes() async throws -> [Athlete]
    func athletes(branchID: EntityID) async throws -> [Athlete]
    func athletes(coachID: EntityID) async throws -> [Athlete]
    func athlete(id: EntityID) async throws -> Athlete?
    func athlete(memberNumber: Int) async throws -> Athlete?
    func nextMemberNumber() async throws -> Int
    func upsert(_ athlete: Athlete) async throws
}

public protocol CoachRepository: Sendable {
    func coaches() async throws -> [Coach]
    func coaches(branchID: EntityID) async throws -> [Coach]
    func coach(id: EntityID) async throws -> Coach?
    func upsert(_ coach: Coach) async throws
}

public protocol ScheduleRepository: Sendable {
    func sessions(branchID: EntityID, on day: Date) async throws -> [ClassSession]
    func sessions(coachID: EntityID, on day: Date) async throws -> [ClassSession]
    func session(id: EntityID) async throws -> ClassSession?
}

public protocol AttendanceRepository: Sendable {
    func attendance(sessionID: EntityID) async throws -> [AttendanceRecord]
    func attendance(athleteID: EntityID, since: Date) async throws -> [AttendanceRecord]
    func upsertAttendance(_ record: AttendanceRecord) async throws
    func upsertAttendance(_ records: [AttendanceRecord]) async throws
}

public protocol PerformanceRepository: Sendable {
    func score(athleteID: EntityID) async throws -> PerformanceScore?
    func scoreHistory(athleteID: EntityID) async throws -> [PerformanceScore]
    func scores(branchID: EntityID) async throws -> [PerformanceScore]
    func allScores() async throws -> [PerformanceScore]
}

public protocol MatchRepository: Sendable {
    func matches(athleteID: EntityID) async throws -> [Match]
    func matches(branchID: EntityID) async throws -> [Match]
    func matches(tournamentID: EntityID) async throws -> [Match]
    func upsertMatch(_ match: Match) async throws
}

public protocol PerformanceEntryRepository: Sendable {
    func physicalTests(athleteID: EntityID) async throws -> [PhysicalTest]
    func assessments(athleteID: EntityID) async throws -> [TechnicalAssessment]
    func wellness(athleteID: EntityID, since: Date) async throws -> [WellnessEntry]
    func upsert(physicalTest: PhysicalTest) async throws
    func upsert(assessment: TechnicalAssessment) async throws
    func upsert(wellness entry: WellnessEntry) async throws
}

public protocol GradingRepository: Sendable {
    func eligibility(athleteID: EntityID, targetBelt: Belt) async throws -> GradingEligibility
    func gradingSessions(branchID: EntityID) async throws -> [GradingSession]
    func gradingSession(id: EntityID) async throws -> GradingSession?
    func upsert(_ session: GradingSession) async throws
    func gradingScores(sessionID: EntityID) async throws -> [GradingScore]
    func upsert(_ score: GradingScore) async throws
    func certificates(athleteID: EntityID) async throws -> [GradingCertificate]
    func issueCertificate(_ certificate: GradingCertificate) async throws
}

public protocol TournamentRepository: Sendable {
    func tournaments() async throws -> [Tournament]
    func tournament(id: EntityID) async throws -> Tournament?
    func upsert(tournament: Tournament) async throws

    func registrations(tournamentID: EntityID) async throws -> [TournamentRegistration]
    func registrations(athleteID: EntityID) async throws -> [TournamentRegistration]
    func upsert(registration: TournamentRegistration) async throws

    func weightCutHistory(registrationID: EntityID) async throws -> [WeightCutEntry]
    func upsert(weightCut: WeightCutEntry) async throws

    func brackets(tournamentID: EntityID) async throws -> [Bracket]
    func upsert(bracket: Bracket) async throws

    func bracketMatches(bracketID: EntityID) async throws -> [BracketMatch]
    func upsert(bracketMatch: BracketMatch) async throws
}

public protocol LiveMatchRepository: Sendable {
    func activeMatch() async throws -> Match?
    func startMatch(_ match: Match) async throws
    func recordEvent(_ event: ScoreEvent) async throws
    func endRound(matchID: EntityID) async throws
    func finalizeMatch(_ match: Match) async throws
    /// AsyncStream of score events for the given match. Demo repository returns
    /// an empty stream; Supabase backend bridges the realtime channel.
    func scoreEventStream(matchID: EntityID) -> AsyncStream<ScoreEvent>
}

public extension LiveMatchRepository {
    func scoreEventStream(matchID: EntityID) -> AsyncStream<ScoreEvent> {
        AsyncStream { $0.finish() }
    }
}

public protocol OperationsRepository: Sendable {
    func announcements(audience: AnnouncementAudience?) async throws -> [Announcement]
    func upsert(announcement: Announcement) async throws
    func rsvps(announcementID: EntityID) async throws -> [AnnouncementRSVP]
    func upsert(rsvp: AnnouncementRSVP) async throws
    func certifications(coachID: EntityID) async throws -> [Certification]
    func certifications() async throws -> [Certification]
    func upsert(certification: Certification) async throws
    func expiringSoon(within: TimeInterval) async throws -> [Certification]
}

public protocol AuditRepository: Sendable {
    func log(_ entry: AuditEntry) async throws
    func entries(actor: EntityID?, since: Date?) async throws -> [AuditEntry]
    func entries(target: EntityID) async throws -> [AuditEntry]
}

public protocol StorageRepository: Sendable {
    /// Uploads athlete photo bytes and returns a publicly-resolvable URL string
    /// the Avatar view can render via AsyncImage.
    func uploadAthletePhoto(athleteID: EntityID, data: Data, contentType: String) async throws -> String
}

public protocol BranchProfileRepository: Sendable {
    func facility(branchID: EntityID) async throws -> BranchFacility?
    func upsert(_ facility: BranchFacility) async throws

    func hours(branchID: EntityID) async throws -> BranchHours?
    func upsert(_ hours: BranchHours) async throws

    func programs(branchID: EntityID) async throws -> [BranchProgram]
    func upsert(_ program: BranchProgram) async throws

    func inventory(branchID: EntityID) async throws -> BranchInventory?
    func upsert(_ inventory: BranchInventory) async throws

    func compliance(branchID: EntityID) async throws -> BranchCompliance?
    func upsert(_ compliance: BranchCompliance) async throws

    func pricing(branchID: EntityID) async throws -> BranchPricing?
    func upsert(_ pricing: BranchPricing) async throws

    func financials(branchID: EntityID, monthsBack: Int) async throws -> [BranchFinancials]
    func upsert(_ financials: BranchFinancials) async throws

    func media(branchID: EntityID) async throws -> BranchMedia?
    func upsert(_ media: BranchMedia) async throws

    func socialLinks(branchID: EntityID) async throws -> BranchSocialLinks?
    func upsert(_ links: BranchSocialLinks) async throws

    func safeguarding(branchID: EntityID) async throws -> BranchSafeguarding?
    func upsert(_ safe: BranchSafeguarding) async throws

    func milestones(branchID: EntityID) async throws -> [BranchMilestone]
    func upsert(_ milestone: BranchMilestone) async throws
}

public protocol Repository:
    UserRepository,
    BranchRepository,
    AthleteRepository,
    CoachRepository,
    ScheduleRepository,
    AttendanceRepository,
    PerformanceRepository,
    MatchRepository,
    PerformanceEntryRepository,
    GradingRepository,
    TournamentRepository,
    LiveMatchRepository,
    OperationsRepository,
    AuditRepository,
    StorageRepository,
    BranchProfileRepository,
    Sendable {}

File: flutter_app/lib/core/repository/repository.dart
import 'dart:typed_data';

import '../models/athlete.dart';
import '../models/athlete_group.dart';
import '../models/audit_log.dart';
import '../models/branch.dart';
import '../models/branch_compliance.dart';
import '../models/branch_facility.dart';
import '../models/branch_financials.dart';
import '../models/branch_hours.dart';
import '../models/branch_inventory.dart';
import '../models/branch_media.dart';
import '../models/branch_milestone.dart';
import '../models/branch_pricing.dart';
import '../models/branch_program.dart';
import '../models/branch_safeguarding.dart';
import '../models/branch_social_links.dart';
import '../models/coach.dart';
import '../models/entity_id.dart';
import '../models/goal.dart';
import '../models/grading.dart';
import '../models/belt.dart';
import '../models/improvement_plan.dart';
import '../models/match.dart';
import '../models/operations.dart';
import '../models/peer_benchmark.dart';
import '../models/performance_entry.dart';
import '../models/performance_score.dart';
import '../models/physical_metric.dart';
import '../models/poomsae_assessment.dart';
import '../models/role.dart';
import '../models/schedule.dart';
import '../models/technical_skill.dart';
import '../models/tournament.dart';
import '../models/training_load_entry.dart';
import '../models/user.dart';

/// Port of the protocol surface in Core/Repository/Repository.swift.
///
/// Each Swift `protocol` maps to one Dart `abstract class`. All 21 sub-protocols
/// are composed into the single [Repository] facade so a single instance
/// (Demo or Supabase) can be injected app-wide via get_it.
///
/// Dart cannot overload by parameter type. Swift overloads are disambiguated
/// with descriptive suffixes — see the disambiguation table in FLUTTER_PORT.md.

// ──────────────────────────────────────────────────────────────────────────────
// 1. UserRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class UserRepository {
  // Existing — must not be renamed (called by session_cubit, screens, tests).
  Future<User?> currentUser();
  Future<List<User>> availableUsers();
  Future<void> setCurrentUser(EntityID id);
  Future<User?> user(EntityID id);

  // New in this port.
  /// Filter by role; null = all roles.
  Future<List<User>> usersByRole(Role? role);
  Future<void> createAccount({
    required String email,
    required String password,
    required String fullName,
    required String fullNameAr,
    required Role role,
    EntityID? branchId,
  });
  Future<void> updateUser(User user);
  Future<void> linkChild({
    required EntityID userId,
    required EntityID athleteId,
  });
  Future<void> unlinkChild({
    required EntityID userId,
    required EntityID athleteId,
  });
}

// ──────────────────────────────────────────────────────────────────────────────
// 2. BranchRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class BranchRepository {
  // Existing.
  Future<List<Branch>> branches();
  Future<Branch?> branch(EntityID id);

  // New.
  Future<void> upsertBranch(Branch branch);
}

// ──────────────────────────────────────────────────────────────────────────────
// 3. AthleteRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class AthleteRepository {
  // Existing — must not be renamed.
  Future<List<Athlete>> athletes();
  Future<List<Athlete>> athletesInBranch(EntityID branchId);
  Future<List<Athlete>> athletesForCoach(EntityID coachId);
  Future<Athlete?> athlete(EntityID id);
  Future<void> upsertAthlete(Athlete athlete);

  // New.
  /// Lookup by member number; Swift: `athlete(memberNumber:)`.
  Future<Athlete?> athleteByMemberNumber(int memberNumber);
  Future<int> nextMemberNumber();
}

// ──────────────────────────────────────────────────────────────────────────────
// 4. CoachRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class CoachRepository {
  Future<List<Coach>> coaches();
  Future<List<Coach>> coachesInBranch(EntityID branchId);
  Future<Coach?> coach(EntityID id);
  Future<void> upsertCoach(Coach coach);
}

// ──────────────────────────────────────────────────────────────────────────────
// 5. ScheduleRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class ScheduleRepository {
  /// Swift: `sessions(branchID:on:)`.
  Future<List<ClassSession>> sessionsForBranch(EntityID branchId, DateTime day);

  /// Swift: `sessions(coachID:on:)`.
  Future<List<ClassSession>> sessionsForCoach(EntityID coachId, DateTime day);

  Future<ClassSession?> session(EntityID id);
  Future<void> upsertSession(ClassSession session);
  Future<void> deleteSession(EntityID id);
}

// ──────────────────────────────────────────────────────────────────────────────
// 6. AttendanceRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class AttendanceRepository {
  /// Swift: `attendance(sessionID:)`.
  Future<List<AttendanceRecord>> attendanceForSession(EntityID sessionId);

  /// Swift: `attendance(athleteID:since:)`.
  Future<List<AttendanceRecord>> attendanceForAthlete(
    EntityID athleteId,
    DateTime since,
  );

  /// Upsert a single record. Swift: `upsertAttendance(_ record:)`.
  Future<void> upsertAttendance(AttendanceRecord record);

  /// Upsert a batch. Swift: `upsertAttendance(_ records:)`.
  Future<void> upsertAttendanceBatch(List<AttendanceRecord> records);
}

// ──────────────────────────────────────────────────────────────────────────────
// 7. PerformanceRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class PerformanceRepository {
  // Existing — must not be renamed.
  Future<PerformanceScore?> score(EntityID athleteId);
  Future<List<PerformanceScore>> scoreHistory(EntityID athleteId);
  Future<List<PerformanceScore>> scoresInBranch(EntityID branchId);
  Future<List<PerformanceScore>> allScores();
}

// ──────────────────────────────────────────────────────────────────────────────
// 8. MatchRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class MatchRepository {
  /// Swift: `matches(athleteID:)`.
  Future<List<Match>> matchesForAthlete(EntityID athleteId);

  /// Swift: `matches(branchID:)`.
  Future<List<Match>> matchesForBranch(EntityID branchId);

  /// Swift: `matches(tournamentID:)`.
  Future<List<Match>> matchesForTournament(EntityID tournamentId);

  Future<void> upsertMatch(Match match);
}

// ──────────────────────────────────────────────────────────────────────────────
// 9. PerformanceEntryRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class PerformanceEntryRepository {
  Future<List<PhysicalMetric>> physicalMetrics(EntityID athleteId);
  Future<List<TechnicalSkill>> technicalSkills(EntityID athleteId);
  Future<List<PoomsaeAssessment>> poomsaeAssessments(EntityID athleteId);

  /// Swift: `wellness(athleteID:since:)`.
  Future<List<WellnessEntry>> wellness(EntityID athleteId, DateTime since);

  /// Swift: `upsert(metric:)`.
  Future<void> upsertPhysicalMetric(PhysicalMetric metric);

  /// Swift: `upsert(skill:)`.
  Future<void> upsertTechnicalSkill(TechnicalSkill skill);

  /// Swift: `upsert(poomsae:)`.
  Future<void> upsertPoomsaeAssessment(PoomsaeAssessment poomsae);

  /// Swift: `upsert(wellness:)`.
  Future<void> upsertWellness(WellnessEntry entry);

  Future<void> deletePhysicalMetric(EntityID id);
  Future<void> deleteTechnicalSkill(EntityID id);
  Future<void> deletePoomsaeAssessment(EntityID id);
}

// ──────────────────────────────────────────────────────────────────────────────
// 10. GoalRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class GoalRepository {
  Future<List<Goal>> goals(EntityID athleteId);

  /// Swift: `upsert(goal:)`.
  Future<void> upsertGoal(Goal goal);
  Future<void> deleteGoal(EntityID id);
}

// ──────────────────────────────────────────────────────────────────────────────
// 11. TrainingLoadRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class TrainingLoadRepository {
  /// Swift: `trainingLoad(athleteID:since:)`.
  Future<List<TrainingLoadEntry>> trainingLoad(
    EntityID athleteId,
    DateTime since,
  );

  /// Swift: `upsert(load:)`.
  Future<void> upsertTrainingLoad(TrainingLoadEntry load);
  Future<void> deleteTrainingLoad(EntityID id);
}

// ──────────────────────────────────────────────────────────────────────────────
// 12. ImprovementPlanRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class ImprovementPlanRepository {
  Future<List<DrillLibraryEntry>> drills();

  /// Swift: `upsert(drill:)`.
  Future<void> upsertDrill(DrillLibraryEntry drill);
  Future<void> deleteDrill(EntityID id);

  Future<List<ImprovementPlan>> improvementPlans(EntityID athleteId);

  /// Swift: `upsert(plan:)`.
  Future<void> upsertPlan(ImprovementPlan plan);
  Future<void> deletePlan(EntityID id);
}

// ──────────────────────────────────────────────────────────────────────────────
// 13. PeerBenchmarkRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class PeerBenchmarkRepository {
  Future<List<PeerBenchmark>> peerBenchmarks();
  Future<void> upsertBenchmarks(List<PeerBenchmark> benchmarks);

  /// Wipe and recompute from current athlete + metric data.
  Future<List<PeerBenchmark>> recomputeBenchmarks();
}

// ──────────────────────────────────────────────────────────────────────────────
// 14. GradingRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class GradingRepository {
  /// Swift: `eligibility(athleteID:targetBelt:)`.
  Future<GradingEligibility> eligibility(EntityID athleteId, Belt targetBelt);

  /// Swift: `gradingSessions(branchID:)`.
  Future<List<GradingSession>> gradingSessionsForBranch(EntityID branchId);

  Future<GradingSession?> gradingSession(EntityID id);

  /// Swift: `upsert(_ session: GradingSession)`.
  Future<void> upsertGradingSession(GradingSession session);

  Future<List<GradingScore>> gradingScores(EntityID sessionId);

  /// Swift: `upsert(_ score: GradingScore)`.
  Future<void> upsertGradingScore(GradingScore gradingScore);

  Future<List<GradingCertificate>> certificates(EntityID athleteId);
  Future<void> issueCertificate(GradingCertificate certificate);
}

// ──────────────────────────────────────────────────────────────────────────────
// 15. TournamentRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class TournamentRepository {
  Future<List<Tournament>> tournaments();
  Future<Tournament?> tournament(EntityID id);

  /// Swift: `upsert(tournament:)`.
  Future<void> upsertTournament(Tournament tournament);

  /// Swift: `registrations(tournamentID:)`.
  Future<List<TournamentRegistration>> registrationsForTournament(
    EntityID tournamentId,
  );

  /// Swift: `registrations(athleteID:)`.
  Future<List<TournamentRegistration>> registrationsForAthlete(
    EntityID athleteId,
  );

  /// Swift: `upsert(registration:)`.
  Future<void> upsertRegistration(TournamentRegistration registration);

  /// Swift: `weightCutHistory(registrationID:)`.
  Future<List<WeightCutEntry>> weightCutHistory(EntityID registrationId);

  /// Swift: `upsert(weightCut:)`.
  Future<void> upsertWeightCut(WeightCutEntry weightCut);

  /// Swift: `brackets(tournamentID:)`.
  Future<List<Bracket>> brackets(EntityID tournamentId);

  /// Swift: `upsert(bracket:)`.
  Future<void> upsertBracket(Bracket bracket);

  /// Swift: `bracketMatches(bracketID:)`.
  Future<List<BracketMatch>> bracketMatches(EntityID bracketId);

  /// Swift: `upsert(bracketMatch:)`.
  Future<void> upsertBracketMatch(BracketMatch bracketMatch);
}

// ──────────────────────────────────────────────────────────────────────────────
// 16. LiveMatchRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class LiveMatchRepository {
  Future<Match?> activeMatch();
  Future<void> startMatch(Match match);
  Future<void> recordEvent(ScoreEvent event);
  Future<void> endRound(EntityID matchId);
  Future<void> finalizeMatch(Match match);

  /// Swift `AsyncStream<ScoreEvent>` → Dart `Stream<ScoreEvent>`.
  /// Demo returns `const Stream.empty()`.
  Stream<ScoreEvent> scoreEventStream(EntityID matchId);
}

// ──────────────────────────────────────────────────────────────────────────────
// 17. OperationsRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class OperationsRepository {
  /// Swift: `announcements(audience:)` — null = all audiences.
  Future<List<Announcement>> announcements(AnnouncementAudience? audience);

  /// Swift: `upsert(announcement:)`.
  Future<void> upsertAnnouncement(Announcement announcement);

  /// Swift: `rsvps(announcementID:)`.
  Future<List<AnnouncementRSVP>> rsvps(EntityID announcementId);

  /// Swift: `upsert(rsvp:)`.
  Future<void> upsertRsvp(AnnouncementRSVP rsvp);

  /// Swift: `certifications(coachID:)`.
  Future<List<Certification>> certificationsForCoach(EntityID coachId);

  /// Swift: `certifications()` — all coaches.
  Future<List<Certification>> certifications();

  /// Swift: `upsert(certification:)`.
  Future<void> upsertCertification(Certification certification);

  /// Swift: `expiringSoon(within:)` — [seconds] maps to Swift's `TimeInterval`.
  Future<List<Certification>> expiringSoon(double seconds);
}

// ──────────────────────────────────────────────────────────────────────────────
// 18. AuditRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class AuditRepository {
  Future<void> log(AuditEntry entry);

  /// Swift: `entries(actor:since:)` — either param may be null.
  Future<List<AuditEntry>> entriesForActor({
    EntityID? actorId,
    DateTime? since,
  });

  /// Swift: `entries(target:)`.
  Future<List<AuditEntry>> entriesForTarget(EntityID targetId);
}

// ──────────────────────────────────────────────────────────────────────────────
// 19. StorageRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class StorageRepository {
  /// Swift: `uploadAthletePhoto(athleteID:data:contentType:)`.
  /// Returns a publicly-resolvable URL string.
  Future<String> uploadAthletePhoto({
    required EntityID athleteId,
    required Uint8List data,
    required String contentType,
  });

  /// Swift: `uploadUserAvatar(userID:data:contentType:)`.
  Future<String> uploadUserAvatar({
    required EntityID userId,
    required Uint8List data,
    required String contentType,
  });
}

// ──────────────────────────────────────────────────────────────────────────────
// 20. BranchProfileRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class BranchProfileRepository {
  Future<BranchFacility?> facility(EntityID branchId);

  /// Swift: `upsert(_ facility:)`.
  Future<void> upsertFacility(BranchFacility facility);

  Future<BranchHours?> hours(EntityID branchId);
  Future<void> upsertHours(BranchHours hours);

  Future<List<BranchProgram>> programs(EntityID branchId);
  Future<void> upsertProgram(BranchProgram program);

  Future<BranchInventory?> inventory(EntityID branchId);
  Future<void> upsertInventory(BranchInventory inventory);

  Future<BranchCompliance?> compliance(EntityID branchId);
  Future<void> upsertCompliance(BranchCompliance compliance);

  Future<BranchPricing?> pricing(EntityID branchId);
  Future<void> upsertPricing(BranchPricing pricing);

  /// Swift: `financials(branchID:monthsBack:)`.
  Future<List<BranchFinancials>> financials(EntityID branchId, int monthsBack);
  Future<void> upsertFinancials(BranchFinancials financials);

  Future<BranchMedia?> media(EntityID branchId);
  Future<void> upsertMedia(BranchMedia media);

  Future<BranchSocialLinks?> socialLinks(EntityID branchId);
  Future<void> upsertSocialLinks(BranchSocialLinks links);

  Future<BranchSafeguarding?> safeguarding(EntityID branchId);
  Future<void> upsertSafeguarding(BranchSafeguarding safe);

  Future<List<BranchMilestone>> milestones(EntityID branchId);
  Future<void> upsertMilestone(BranchMilestone milestone);
}

// ──────────────────────────────────────────────────────────────────────────────
// 21. AthleteGroupRepository
// ──────────────────────────────────────────────────────────────────────────────

abstract class AthleteGroupRepository {
  Future<List<AthleteGroup>> athleteGroups();
  Future<AthleteGroup?> athleteGroup(EntityID id);

  /// Swift: `upsert(_ group:)`.
  Future<void> upsertGroup(AthleteGroup group);
  Future<void> deleteAthleteGroup(EntityID id);
}

// ──────────────────────────────────────────────────────────────────────────────
// Composed facade
// ──────────────────────────────────────────────────────────────────────────────

/// The composed facade injected app-wide. Mirrors the Swift `Repository`
/// protocol-composition pattern (21 sub-protocols). Repository swap is one
/// line in `lib/app/locator.dart`.
abstract class Repository
    implements
        UserRepository,
        BranchRepository,
        AthleteRepository,
        CoachRepository,
        ScheduleRepository,
        AttendanceRepository,
        PerformanceRepository,
        MatchRepository,
        PerformanceEntryRepository,
        GoalRepository,
        TrainingLoadRepository,
        ImprovementPlanRepository,
        PeerBenchmarkRepository,
        GradingRepository,
        TournamentRepository,
        LiveMatchRepository,
        OperationsRepository,
        AuditRepository,
        StorageRepository,
        BranchProfileRepository,
        AthleteGroupRepository {}

/// Optional auth surface — mirrors the Swift `AuthenticatingRepository`
/// protocol (separate from [Repository] because not every backend
/// authenticates). [DemoRepository] implements it as a no-friction demo
/// switch; [SupabaseRepository] backs it with real Supabase Auth. Callers
/// feature-detect via `repo is AuthRepository`.
abstract class AuthRepository {
  Future<void> signInWithEmail({required String email, required String password});
  Future<void> signOut();
}

/// Thrown when account creation or update targets the reserved [AppOwner]
/// email on a non-owner record — the owner account can't be created,
/// impersonated, demoted, or renamed by any other user.
class OwnerEmailReservedException implements Exception {
  const OwnerEmailReservedException();
  @override
  String toString() => 'That email address is reserved.';
}
